harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+        # checkpoints are needed.
+        # TODO (Liam): revisit later to optimize sharded checkpoint loading.
+
+        # Load stateful things tracked by Determined on all slots.
+        checkpoint: Optional[Dict[str, Any]] = None
+        ckpt_path = "det_state_dict.pth"
+        maybe_ckpt = load_path.joinpath(ckpt_path)
+        if maybe_ckpt.exists():
+            checkpoint = torch.load(str(maybe_ckpt), map_location="cpu")  # type: ignore
+        if checkpoint is None or not isinstance(checkpoint, dict):
+            return
+
+        for callback in self.callbacks.values():
+            callback.on_checkpoint_load_start(checkpoint)
+
+        # We allow users to override load behavior if needed but we default to using
+        # the load method provided by DeepSpeed.
+        self.trial.load(self.context, load_path)
+
+        if "rng_state" in checkpoint:
+            rng_state = checkpoint["rng_state"]
+            np.random.set_state(rng_state["np_rng_state"])
+            random.setstate(rng_state["random_rng_state"])
+            torch.random.set_rng_state(rng_state["cpu_rng_state"])
+
+            if torch.cuda.device_count():
+                if "gpu_rng_state" in rng_state:
+                    torch.cuda.set_rng_state(
+                        rng_state["gpu_rng_state"], device=self.context.distributed.get_local_rank()
+                    )
+                else:
+                    logging.warning(
+                        "The system has a gpu but no gpu_rng_state exists in the checkpoint."
+                    )
+            else:
+                if "gpu_rng_state" in rng_state:
+                    logging.warning(
+                        "There exists gpu_rng_state in checkpoint but the system has no gpu."
+                    )
+        else:
+            logging.warning("The checkpoint has no random state to restore.")
+
+        callback_state = checkpoint.get("callbacks", {})
+        for name in self.callbacks:
+            if name in callback_state:
+                self.callbacks[name].load_state_dict(callback_state[name])
+            elif util.is_overridden(self.callbacks[name].load_state_dict, pytorch.PyTorchCallback):
+                logging.warning(
+                    "Callback '{}' implements load_state_dict(), but no callback state "
+                    "was found for that name when restoring from checkpoint. This "
+                    "callback will be initialized from scratch"
+                )
+
+        # Load workload sequencer state.
+        wlsq_path = load_path.joinpath("workload_sequencer.pkl")
+        if self.wlsq is not None and wlsq_path.exists():
+            with wlsq_path.open("rb") as f:
+                self.wlsq.load_state(pickle.load(f))
+
+    def _save(self, path: pathlib.Path) -> None:
+        if self.context.distributed._is_local_chief:
+            path.mkdir(parents=True, exist_ok=True)
+        self.context.distributed._zmq_gather(None)  # sync
```

Comment:
maybe `_zmq_gather_local`?  Global sync seems unnecessary here.

Also, nit: can you do `_ = self.context....` to explicitly indicate you are ignoring a return value?